Phil/activate moar

psFried opened the pull request:
**Description:**

Resolves #2170 
Resolves #2171

Introduces some major changes to how controllers keep tabs on task shards. Previously, controllers would always assume that any successful activation would mean that we'd be guaranteed to see a ShardFailed event if it failed again. But this turned out to be something we can't easily guarantee. Further, we can't ever get ShardFailed events for shards of the ops catalog itself, and so we need to rely on another mechanism for monitoring those shards.

The solution in both cases is for controller to periodically do shard listings in order to keep track of the status of task shards.

### Controller changes

Controllers will periodically check the status of task shards. There's a new `shard_status` field in the controller status, which tracks the results of those checks.

Controllers will use an increasingly long duration between shard status checks as long as the status hasn't changed. We want to avoid having controllers running too frequently or sending too many requests to reactors. A change in the observed shard status resets this duration, so that we'll check more frequently right after a change. Also, we use much longer durations when shards are healthy since we should be receiving ShardFailed events if there's a failure after a shard reaches Primary status.

Controllers will attempt to re-activate task shards if:
- any shards are Failed for 3 consecutive checks
- it's been at least 15 minutes (configurable via `FLOW_REACTIVATE_INTERVAL` env variable) since the last time shards were activated
- No `ShardFailed` event was observed

In scenarios where we _do_ get a ShardFailed event, the above logic will not apply, and the current behavior remains in place. This allows us to continue restarting failed shards immediately after the first couple of failures, with an incremental backoff after multiple failures.

In scenarios where we _don't_ get a ShardFailed event, it means that failed shards will be re-activated after 3 consecutive failed status checks (which takes ~2 minutes in total after the first failed check), and at most once every 15 minutes.

### Ops catalog shards

There's a special carve out for ops catalog tasks so that we check their shard statuses much more frequently, and re-start them more quickly if they're failed. Apart from that, ops catalog shards are covered by the same controller behavior that's used for all other tasks. Once this is deployed, we should be able to remove all the shard-supervisor cron jobs, as controllers will be able to handle all shard re-starts.

### Status summaries

I also updated how we summarize status information, so that the summary now accounts for shard failures that were observed via periodic shard listings. The new integration tests use the status summaries pretty extensively, so test coverage of status summaries has increased quite a bit, and even surfaced a few bugs that are fixed here.

**Workflow steps:**

No user-facing changes, except for the output of `flowctl catalog status`, which hopefully just got a bit more helpful and correct.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/flow/2182)
<!-- Reviewable:end -->

psFried committed:
agent: controllers check shard statuses

Fixes #2170

Updates controllers to periodically check shard statuses. Controllers store
shard status summary in new `ShardStatusCheck`, and the `/catalog/status`
endpoint will use this status when generating the `summary`.

Controllers will also use shard status information in order to periodically
re-activate failed shards, even when no `ShardFailed` events are observed. This
ensures that shards get restarted even when there's a failure in the ops
catalog or if the data plane is unable to write a `ShardFailed` event.

The `shard_health` only applies to enabled tasks that have task shards (not
dekaf tasks).

Also updates the status summary to only expect a connector status for tasks
that have previously had a connector status present. This fixes an issue where
we previously would show `waiting on connector health check` as the status
summary for all dekaf tasks. This approach allows connectors to use or ignore
the connector status, at the expense of adding in a (hopefully trivial) edge
case where a newly created task will show as `Ok` before a connector status is
observed (only after shards are `Primary` though).

psFried committed:
agent: controllers handle ops catalog shard failures

Adds a special case to controllers for tasks that are part of the ops catalog.
Because these tasks are involved in materializing the `ShardFailed` events, we
can't rely on the delivery of those events in order to restart ops catalog
shards.  This introduces a special case to have controllers do more frequent
health checks of ops catalog shards, and to re-activate the shards much more
quickly than we would for normal tasks.

psFried committed:
agent: fix linter warning on App visibility

Changes the visibility of `App` and some functions that use it to be
`pub(crate)`. This just quiets a few warnings about private types being exposed
on public functions.

psFried committed:
agent: more comprehensive shard failures tests

Moves the shard failures integration tests into a separate module, and make
them much more comprehensive. The tests now cover all task types, and also
cover the interplay between ShardFailed events and periodic checks of the shard
statuses.

This includes fixes for a few bugs I discovered along the way:

- Catalog tests would show as failing if the controller ran and didn't need to publish.
- Status summary would be TaskDisabled for disabled derivations, and should instead be Ok.

psFried committed:
agent: log warning if shards are pending for too long

Adds a warning log so that we can tell if shards are taking too long to come
up. The intent is to setup an alert if we start seeing too many of these.

jgraettinger commented on crates/agent/src/controllers/activation.rs:
```diff
@@ -115,23 +193,263 @@ pub async fn update_activation<C: ControlPlane>(
+        .await
+        .context("listing task shards")?;
+
+    if list_response.status != proto_gazette::consumer::Status::Ok as i32 {
```
the gazette client already checks status and maps into an error:
https://github.com/estuary/flow/blob/master/crates/gazette/src/shard/mod.rs#L60

jgraettinger commented on crates/agent/src/controllers/activation.rs:
```diff
@@ -115,23 +193,263 @@ pub async fn update_activation<C: ControlPlane>(
+        // This can happen if the ops catalog is broken, or in certain edge
+        // cases in the data plane.
+        tracing::warn!(%time_since_activation, failed_checks = %count, "re-activating task shards because they still show as Failed after prior activation");
+        do_activate(now, state, status, control_plane).await?;
```
discussion: a little hand-wavy, but you could have it send itself a ShardFailed event which would keep a single control flow for how the controller "observes" a failure event. Not asserting that's worth it / makes sense but it's an intuitive recursive framing to me at least. 

The flow in here is complex (but so is the domain), so i'm just casting about for any patterns which might help simplify.

psFried replied:
I did consider that, but only until it started getting a bit murky. I was uncertain about how to map an observation of Failed shards via listing into `ShardFailed` events, and saw a couple of challenges:

What if the actual `ShardFailed` event just gets materialized after we've observed the failed shards via listing? A slowdown in the ops catalog could result in duplicate `ShardFailed` events, which would result in longer restart backoffs. If there was a good way to deduplicate those ShardFailed events, we'd be all set, but I can't think of one. Definitely open to suggestions here, because I'm with you on wanting to reign in the complexity if we can.

jgraettinger replied:
> What if the actual ShardFailed event just gets materialized after we've observed the failed shards via listing? A slowdown in the ops catalog could result in duplicate ShardFailed events, which would result in longer restart backoffs.

I don't think this is unique to this question -- for example, a reactor deployment will produce extra ShardFailed events just due to the elevated hand-offs.

psFried replied:
Then I'll take a crack at it, and see how it goes. Will update here once I try it out

jgraettinger reviewed (approved):
LGTM

psFried commented on crates/agent/src/controllers/activation.rs:
```diff
@@ -115,23 +193,263 @@ pub async fn update_activation<C: ControlPlane>(
+
+    // If there's been at least 3 failed checks in a row, and it's been longer than the
+    // re-activation interval, then re-activate failed shards. We require at least two
+    // failed checks in a row because sometimes shards can temporarily be in failed state
```
Yeah, this is because we'd rather have the normal shard restart mechanism handle these, but there's a race condition where we could observe the failure via a shard listing before the `ShardFailed` event gets materialized. So the idea was to give it a couple of minutes before acting on the this.

psFried committed:
agent: synthesize ShardFailed events from failed status checks

Updates the activation controller to change how it responds to failed status
checks.  The controller will now insert synthetic ShardFailed events after
observing a shard failure via periodic status checks. This just simplifies the
code so that there's only one code path for handling backoffs and
re-activations.

jgraettinger reviewed (approved):
still LGTM % nit
